provider/oci/images_test.go

```diff
@@ -0,0 +1,393 @@
+		OperatingSystemVersion: makeStringPointer("14.04"),
+		DisplayName:            makeStringPointer("Canonical-Ubuntu-14.04-2018.01.11-0"),
+	}
+
+	imgType, err := oci.NewInstanceImage(image, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Check(imgType.ImageType, gc.Equals, oci.ImageTypeGeneric)
+	c.Check(imgType.Series, gc.Equals, "trusty")
+	c.Check(imgType.CompartmentId, gc.NotNil)
+	c.Check(*imgType.CompartmentId, gc.Equals, i.testCompartment)
+	c.Check(imgType.Id, gc.Equals, i.testImageID)
+}
+
+func (i *imagesSuite) TestNewInstanceImageUnknownOS(c *gc.C) {
+	image := ociCore.Image{
+		CompartmentId:          &i.testCompartment,
+		Id:                     &i.testImageID,
+		OperatingSystem:        makeStringPointer("NotKnownToJuju"),
+		OperatingSystemVersion: makeStringPointer("14.04"),
+		DisplayName:            makeStringPointer("Canonical-Ubuntu-14.04-2018.01.11-0"),
+	}
+
+	_, err := oci.NewInstanceImage(image, &i.testCompartment)
+	c.Assert(err, gc.ErrorMatches, "os NotKnownToJuju not supported")
+}
+
+func (i *imagesSuite) TestRefreshImageCache(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	compute := ocitesting.NewMockOCIComputeClient(ctrl)
+	defer ctrl.Finish()
+
+	fakeUbuntuIDSecond := "fakeUbuntu2"
+
+	listImageRequest, listImageResponse := makeListImageRequestResponse(
+		i.testCompartment, []map[string]string{
+			map[string]string{
+				"id":      "fakeUbuntu1",
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.11-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      fakeUbuntuIDSecond,
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.12-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      "fakeCentOS",
+				"os":      "CentOS",
+				"name":    "CentOS-7-2017.10.19-0",
+				"version": "7",
+			},
+		})
+
+	shapesRequestUbuntu, shapesResponseUbuntu := makeShapesRequestResponse(
+		i.testCompartment, "fakeUbuntu1", []string{"VM.Standard2.1", "VM.Standard1.2"})
+
+	shapesRequestUbuntuSecond := ociCore.ListShapesRequest{
+		CompartmentId: &i.testCompartment,
+		ImageId:       &fakeUbuntuIDSecond,
+	}
+
+	shapesRequestCentOS, shapesResponseCentOS := makeShapesRequestResponse(
+		i.testCompartment, "fakeCentOS", []string{"VM.Standard1.2"})
+
+	gomock.InOrder(
+		compute.EXPECT().ListImages(context.Background(), listImageRequest).Return(listImageResponse, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntu).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntuSecond).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestCentOS).Return(shapesResponseCentOS, nil),
+	)
+
+	imgCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Assert(imgCache, gc.NotNil)
+	c.Check(len(imgCache.ImageMap()), gc.Equals, 2)
```

Comment:
Do you know about
c.Check(imgCache.ImageMap(), gc.HasLen, 2)

it depends whether on failure you would like to see "1 != 2" or whether you would like to see:
[foo bar] does not have length 2

ImageMap may be involved enough that it is cleaner to do the len() check, but I think for some of the others, gc.HasLen gives errors that give you a bit more context as to what is actually wrong.

Reply:
Didn't know about ```gc.HasLen```. Thanks!